LibreNMS/service.py

```diff
@@ -0,0 +1,516 @@
+                                                            self.dispatch_discovery)
+        self.billing_manager = LibreNMS.BillingQueueManager(self.config, self.poll_billing,
+                                                            self.dispatch_poll_billing, self.dispatch_calculate_billing)
+
+        self.daily_timer.start()
+
+        print("LibreNMS Service: {} started!".format(self.config.unique_name))
+        print("Poller group {}. Using Python {} and {} locks and queues"
+              .format('0 (default)' if self.config.group == 0 else self.config.group, python_version(),
+                      'redis' if isinstance(self._lm, LibreNMS.RedisLock) else 'internal'))
+        info("Maintenance tasks will be run every {}".format(timedelta(seconds=self.config.update_frequency)))
+
+        # Main dispatcher loop
+        try:
+            while True:
+                master_lock = self._lm.lock('dispatch.master', self.config.unique_name, self.config.master_timeout, True)
+                if master_lock:
+                    if not self.is_master:
+                        info("{} is now the master dispatcher".format(self.config.name))
+                        self.is_master = True
+                        self.start_dispatch_timers()
+
+                    devices = self.fetch_immediate_device_list()
+                    for device in devices:
+                        device_id = device[0]
+                        group = device[1]
+
+                        if device[2]:  # polling
+                            self.dispatch_immediate_polling(device_id, group)
+
+                        if device[3]:  # discovery
+                            self.dispatch_immediate_discovery(device_id, group)
+                else:
+                    if self.is_master:
+                        info("{} is no longer the master dispatcher".format(self.config.name))
+                        self.stop_dispatch_timers()
+                        self.is_master = False  # no longer master
+                sleep(self.config.master_resolution)
+        except KeyboardInterrupt:
+            pass
+
+        self.shutdown()
+
+    # ------------ Discovery ------------
+    def dispatch_immediate_discovery(self, device_id, group):
+        if self.discovery_manager.get_queue(group).empty() and not self.discovery_is_locked(device_id):
+            self.discovery_manager.post_work(device_id, group)
+
+    def dispatch_discovery(self):
+        devices = self.fetch_device_list()
+        for device in devices:
+            self.discovery_manager.post_work(device[0], device[1])
+
+    def discover_device(self, device_id):
+        if self.lock_discovery(device_id):
+            try:
+                info("Discovering device {}".format(device_id))
+                self.call_script('discovery.php', ('-h', device_id))
+                info('Discovery complete {}'.format(device_id))
+            except subprocess.CalledProcessError as e:
```

Comment:
This doesn't appear to be firing - ran out of time to pin down the root cause, but it seems that discovery.php is returning 0 when failing to discover under some circumstance:

```
[root@sdc-mon-001 librenms]# ./discovery.php -d -h 118
LibreNMS Discovery
===================================
Version info:
Commit SHA: faf48311b2c919d9bdde49220536cd0cb95f7930
Commit Date: 1521744011
DB Schema: 247
PHP: 7.1.14
MySQL: 5.5.56-MariaDB
RRDTool: 1.4.8
SNMP: NET-SNMP 5.7.2
==================================DEBUG!
Updating os_def.cache... Done
SQL[SELECT * FROM `devices` WHERE disabled = 0 AND snmp_disable = 0 AND `device_id` = '118' AND poller_group IN(1,3,2,4) ORDER BY device_id DESC]
SQL[SELECT * FROM devices_attribs WHERE `device_id` = '118']
ndc-mon-001.inf.virt.ja.net 118 linux SQL[INSERT INTO `device_perf` (`xmt`,`rcv`,`loss`,`min`,`max`,`avg`,`device_id`,`timestamp`)  VALUES ('3','0','100','0','0','0','118',NOW())]
Unpingable./discovery.php 118 2018-03-22 20:42:31 - 0 devices discovered in 1.958 secs
SNMP [0/0.00s]: Get[0/0.00s] Getnext[0/0.00s] Walk[0/0.00s]
MySQL [7/0.01s]: Cell[2/0.00s] Row[0/-0.00s] Rows[4/0.00s] Column[0/0.00s] Update[0/0.00s] Insert[1/0.00s] Delete[0/0.00s]
RRD [0/0.00s]: Update[0/0.00s] Create [0/0.00s] Other[0/0.00s]
[root@sdc-mon-001 librenms]# echo $?
0
```